<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Basic usage · IdentityByDescentDispersal.jl</title><meta name="title" content="Basic usage · IdentityByDescentDispersal.jl"/><meta property="og:title" content="Basic usage · IdentityByDescentDispersal.jl"/><meta property="twitter:title" content="Basic usage · IdentityByDescentDispersal.jl"/><meta name="description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="og:description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="twitter:description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="og:url" content="https://currocam.github.io/IdentityByDescentDispersal.jl/tutorial/"/><meta property="twitter:url" content="https://currocam.github.io/IdentityByDescentDispersal.jl/tutorial/"/><link rel="canonical" href="https://currocam.github.io/IdentityByDescentDispersal.jl/tutorial/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">IdentityByDescentDispersal.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../overview/">Theory overview</a></li><li class="is-active"><a class="tocitem" href>Basic usage</a><ul class="internal"><li><a class="tocitem" href="#Expected-density-of-identity-by-descent-blocks"><span>Expected density of identity-by-descent blocks</span></a></li><li><a class="tocitem" href="#Inference"><span>Inference</span></a></li></ul></li><li><a class="tocitem" href="../inference/">Inference and model evaluation</a></li><li><span class="tocitem">Developer docs</span><ul><li><a class="tocitem" href="../95-reference/">Reference</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Basic usage</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Basic usage</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/main/docs/src/tutorial.jl#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><p>This section demonstrates the basic usage of the package. We refer to the <code>Inference and model evaluation</code> section of the documentation for an explanation of how to use this package to estimate parameters of demographic models from observed data.</p><pre><code class="language-julia hljs">using IdentityByDescentDispersal</code></pre><h2 id="Expected-density-of-identity-by-descent-blocks"><a class="docs-heading-anchor" href="#Expected-density-of-identity-by-descent-blocks">Expected density of identity-by-descent blocks</a><a id="Expected-density-of-identity-by-descent-blocks-1"></a><a class="docs-heading-anchor-permalink" href="#Expected-density-of-identity-by-descent-blocks" title="Permalink"></a></h2><p>Genetic distances between populations are often a function of geographic distance. This phenomenon is often referred to as isolation-by-distance. Isolation-by-distance is reflected in the patterns of identity-by-descent blocks (also known as IBD blocks). It is possible to derive the expected patterns of shared identity-by-descent blocks under a diffusion approximation and certain demographic models. We refer to the publication of Ringbauer et al. (2017) and the <code>Model overview</code> section of the documentation for more details.</p><p>The main feature of this package is to compute the expected density of identity-by-descent blocks under different demographic scenarios. Next, we show how to compute the expected density of identity-by-descent blocks under a constant population density using the <code>expected_ibd_blocks_constant_density</code> function.</p><pre><code class="language-julia hljs">using Plots

L = 0.01         # Block length threshold (in Morgans)
G = 1.0          # Genome length (in Morgans)
r_values = range(0.01, 25.0, length = 200);  # Distances

plot(
    r_values,
    expected_ibd_blocks_constant_density.(r_values, 1.0, 0.5, L, G),
    xlabel = &quot;Geographic distance between individuals (r)&quot;,
    ylabel = &quot;E[#IBD blocks per unit of block length and per pair]&quot;,
    title = &quot;Constant effective density scenario&quot;,
    label = &quot;D=1.0, σ=0.5&quot;,
)
plot!(
    r_values,
    expected_ibd_blocks_constant_density.(r_values, 2.0, 0.5, L, G),
    label = &quot;D=2.0, σ=0.5&quot;,
)
plot!(
    r_values,
    expected_ibd_blocks_constant_density.(r_values, 1.0, 0.8, L, G),
    label = &quot;D=1.0, σ=0.8&quot;,
)</code></pre><img src="781f8d42.svg" alt="Example block output"/><p>The figure above shows how the density of IBD blocks of length 1 centimorgan in a genome of 1 Morgan decays as a function of geographic distance for different demographic scenarios that vary in effective density and dispersal rate. The effective density is in units of number of diploid individuals per area unit (e.g., km²). The dispersal rate is defined as the square root of the average squared axial parent-offspring distance. The rate of the decay of identity-by-descent blocks is directly related to the time since the most recent common ancestor.</p><p>Recall that the length of IBD blocks is a continuous random variable. Therefore, the expected number of IBD blocks of exactly length <code>L</code> is zero. What the <code>expected_ibd_blocks_constant_density</code> function returns is the expected number of IBD blocks of length <code>L</code> per unit of block length (in Morgans) and per pair. In order to compare such expectations with observed data, we often want to consider instead the expected number of blocks whose length falls in a given interval <code>[a, b]</code>. For a small enough interval <span>$[L, L + \Delta L]$</span> we can approximate the expected number of blocks whose length falls in that interval as:</p><p class="math-container">\[E[N_{[L, L + \Delta L]}] = \int_{L}^{L + \Delta L} E[N_L] dL \approx E[N_L] \Delta L\]</p><p>For example, the expected number of identity-by-descent blocks whose length falls in the interval <code>[1cM, 1.1cM]</code> shared by two diploid individuals that are 1 kM apart in a population with constant effective population density of <code>10/kM²</code>and dispersal rate of <code>0.5/kM</code> would be</p><pre><code class="language-julia hljs">expected_ibd_blocks_constant_density(1.0, 10.0, 0.5, L, G) * 0.001</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">0.6299519796748007</code></pre><p>Ringbauer et al. (2017) derived analytical solutions for the expected density of identity-by-descent blocks for the family of functions of effective population density of the form <span>$D_e(t) = D_1 t^{-\beta}$</span>. A constant effective population density is a special case of this family with <span>$\beta = 0$</span>. For other values of <span>$\beta$</span>, the expected density of identity-by-descent blocks can be computed using the <code>expected_ibd_blocks_power_density</code> function.</p><p>In addition to power density functions, we also support arbitrary effective density functions provided by the user via numerical integration. Next, we illustrate how to define a custom function by considering the popular choice of an exponential growth function.</p><pre><code class="language-julia hljs">D = 1.0          # Effective population density
σ = 1.0          # Dispersal rate
function De(t, θ)
    D₀, α = θ
    max(D₀ * exp(-α * t), eps())
end
plot(
    r_values,
    [expected_ibd_blocks_custom(r, De, [D, 0.0], σ, L, G) for r in r_values],
    xlabel = &quot;Geographic distance between individuals (r)&quot;,
    ylabel = &quot;E[#IBD blocks per unit of block length and per pair]&quot;,
    title = &quot;Constant effective density scenario&quot;,
    label = &quot;growth rate α = 0.0&quot;,
)
plot!(
    r_values,
    [expected_ibd_blocks_custom(r, De, [D, 0.005], σ, L, G) for r in r_values],
    label = &quot;growth rate α = 0.005&quot;,
)
plot!(
    r_values,
    [expected_ibd_blocks_custom(r, De, [D, -0.005], σ, L, G) for r in r_values],
    label = &quot;growth rate α = -0.005&quot;,
)</code></pre><img src="93f18445.svg" alt="Example block output"/><p>It is possible to define more complex effective density functions. However, users should consider carefully whether such parameters are identifiable from identity-by-descent blocks if they aim to estimate them. As an example, we compare the one scenario with an oscillating effective density function with a constant effective density demography.</p><pre><code class="language-julia hljs">function De(t, θ)
    D₀, a, ω = θ
    D₀ * (1 + a * sin(ω * t))
end
D = 1.0          # Effective population density
σ = 1.0          # Dispersal rate
θ = [D, 0.5, 2π]  # Parameters for De(t): D₀, a, ω
t_values = range(0.0, 10.0, length = 200)   # Time for plotting D_e(t)
p1 = plot(
    t_values,
    [De(t, θ) for t in t_values],
    xlabel = &quot;Time (generations ago)&quot;,
    ylabel = &quot;Effective population density&quot;,
    label = &quot;Oscillating effective density&quot;,
)
hline!(p1, [D], label = &quot;Constant effective density&quot;)
p2 = plot(
    r_values,
    [expected_ibd_blocks_custom(r, De, θ, σ, L, G) for r in r_values],
    xlabel = &quot;Geographic distance between individuals (r)&quot;,
    ylabel = &quot;E[#IBD blocks per unit of block length and per pair]&quot;,
    label = &quot;Oscillating effective density&quot;,
)
plot!(
    p2,
    r_values,
    [expected_ibd_blocks_constant_density(r, D, σ, L, G) for r in r_values],
    label = &quot;Constant effective density&quot;,
)

plot(p1, p2, layout = (2, 1), size = (600, 800))</code></pre><img src="95ffd7c8.svg" alt="Example block output"/><h2 id="Inference"><a class="docs-heading-anchor" href="#Inference">Inference</a><a id="Inference-1"></a><a class="docs-heading-anchor-permalink" href="#Inference" title="Permalink"></a></h2><p>Ringbauer et. al (2017) proposed to do inference by assuming that the observed number of IBD blocks that a pair <span>$r$</span> units apart that fall within a small bin <span>$[L_i, L_i + \Delta L]$</span> follows a Poisson distribution with mean <span>$E[N_{L_i}(r, \theta)] \Delta L$</span>.</p><p>Therefore, the log probability of observing <code>Y</code> identity-by-descent blocks whose length fall in the bin <span>$[L_i, L_i + \Delta L]$</span> from a pair of individuals that are <span>$r$</span> units apart can be calculated simply as:</p><pre><code class="language-julia hljs">using Distributions
λ = expected_ibd_blocks_constant_density(
    0.2, # Distance,
    0.5, # Dispersal rate
    2.0, # Effective density
    0.01, # Block length threshold (in Morgans)
    1.0, # Genome length (in Morgans)
)
Y = 5 # Observed number of IBD blocks
logpdf(Poisson(λ), Y)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">-775.0267763054296</code></pre><p>The inference scheme is based on a composite-likelihood that treats each bin from different pairs of individuals as independent. For computational reasons, it is also faster to aggregate observations that correspond to the same distance and bin.</p><p>The input data for this sort of analysis is:</p><ul><li>A <code>DataFrame</code> containing the length of identity-by-descent blocks shared across different individuals.</li><li>A <code>DataFrame</code> containing distances across pairs of individuals.</li><li>A list of contig lengths to properly take into account chromosomal edges (in Morgans).</li></ul><pre><code class="language-julia hljs">using DataFrames
ibd_blocks = DataFrame(
    ID1 = [&quot;A&quot;, &quot;B&quot;, &quot;A&quot;, &quot;C&quot;, &quot;B&quot;],
    ID2 = [&quot;B&quot;, &quot;A&quot;, &quot;C&quot;, &quot;A&quot;, &quot;C&quot;],
    span = [0.005, 0.012, 0.21, 0.10, 0.08], # Morgans
)
individual_distances = DataFrame(
    ID1 = [&quot;A&quot;, &quot;A&quot;, &quot;B&quot;],
    ID2 = [&quot;B&quot;, &quot;C&quot;, &quot;C&quot;],
    distance = [10.0, 20.0, 10.0], # (e.g., in kilometers)
)
contig_lengths = [1.0, 1.5]; # Contig lengths (in Morgans)</code></pre><p>Preprocessing the data requires combining the different sources of information and binning the identity-by-descent blocks. For this purpose, we provide a helper function <code>preprocess_dataset</code> that returns a <code>DataFrame</code> in a &quot;long&quot; format. Of course, you may specify your own bins, but here we use the same bins used by Ringbauer et al. (2017), which can be accessed via the <code>default_ibd_bins</code> function.</p><pre><code class="language-julia hljs">bins, min_length = default_ibd_bins()
df = preprocess_dataset(ibd_blocks, individual_distances, bins, min_length);</code></pre><p>In most scenarios, distances between diploid individuals are not available, but between sampling sites. The <code>preprocess_dataset</code> function handles this scenario appropriately by aggregating observations from pairs that have the exact same distance. However, if distances are available at the individual level, you may consider binning them to reduce the runtime. You might find an example of this in the <a href="https://github.com/currocam/IdentityByDescentDispersal.jl/tree/main/simulations">simulations</a> subdirectory.</p><p>We can now use the df <code>DataFrame</code>to compute composite log likelihoods of different parameters using the family of <code>composite_loglikelihood_*</code> functions. The second major feature of this package is that functions are all compatible with automatic differentiation. Analytical solutions depend on a modified Bessel function of the second kind, which is not widely implemented in a way that allows for automatic differentiation . Here, we rely on the <code>BesselK.jl</code> package for doing this. In addition, we rely on the system of <code>QuadGK.jl</code> for numerical integration compatible with automatic differentiation. This allows us to perform gradient-based optimization and interact with existing software.</p><p>We recommend using this package together with the <code>Turing.jl</code>, as it provides a convenient interface for Bayesian and frequentist inference.</p><h3 id="Maximum-likelihood-estimation"><a class="docs-heading-anchor" href="#Maximum-likelihood-estimation">Maximum likelihood estimation</a><a id="Maximum-likelihood-estimation-1"></a><a class="docs-heading-anchor-permalink" href="#Maximum-likelihood-estimation" title="Permalink"></a></h3><p>For example, we can estimate the parameters of a constant density model using maximum likelihood estimation:</p><pre><code class="language-julia hljs">using Turing, StatsBase, StatsPlots
@model function constant_density(df, contig_lengths)
    D ~ Uniform(0, 100)
    σ ~ Uniform(0, 20)
    Turing.@addlogprob! composite_loglikelihood_constant_density(D, σ, df, contig_lengths)
end</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">constant_density (generic function with 2 methods)</code></pre><p>Generate a MLE estimate.</p><pre><code class="language-julia hljs">mle_estimate = maximum_likelihood(constant_density(df, contig_lengths))
DataFrame(coeftable(mle_estimate)) # computed from the Fisher information matrix</code></pre><div><div style = "float: left;"><span>2×7 DataFrame</span></div><div style = "clear: both;"></div></div><div class = "data-frame" style = "overflow-x: scroll;"><table class = "data-frame" style = "margin-bottom: 6px;"><thead><tr class = "columnLabelRow"><th class = "stubheadLabel" style = "font-weight: bold; text-align: right;">Row</th><th style = "text-align: left;">Name</th><th style = "text-align: left;">Coef.</th><th style = "text-align: left;">Std. Error</th><th style = "text-align: left;">z</th><th style = "text-align: left;">Pr(&gt;|z|)</th><th style = "text-align: left;">Lower 95%</th><th style = "text-align: left;">Upper 95%</th></tr><tr class = "columnLabelRow"><th class = "stubheadLabel" style = "font-weight: bold; text-align: right;"></th><th title = "String" style = "text-align: left;">String</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th></tr></thead><tbody><tr class = "dataRow"><td class = "rowLabel" style = "font-weight: bold; text-align: right;">1</td><td style = "text-align: left;">D</td><td style = "text-align: right;">0.0313893</td><td style = "text-align: right;">0.179795</td><td style = "text-align: right;">0.174583</td><td style = "text-align: right;">0.861407</td><td style = "text-align: right;">-0.321003</td><td style = "text-align: right;">0.383781</td></tr><tr class = "dataRow"><td class = "rowLabel" style = "font-weight: bold; text-align: right;">2</td><td style = "text-align: left;">σ</td><td style = "text-align: right;">20.0</td><td style = "text-align: right;">58.1803</td><td style = "text-align: right;">0.343759</td><td style = "text-align: right;">0.731027</td><td style = "text-align: right;">-94.0313</td><td style = "text-align: right;">134.031</td></tr></tbody></table></div><h3 id="Bayesian-inference"><a class="docs-heading-anchor" href="#Bayesian-inference">Bayesian inference</a><a id="Bayesian-inference-1"></a><a class="docs-heading-anchor-permalink" href="#Bayesian-inference" title="Permalink"></a></h3><p>Alternatively, we can do a standard Bayesian inference with any of the available inference algorithms such as NUTS. Here, we fit a power-density model using NUTS:</p><pre><code class="language-julia hljs">@model function power_density(df, contig_lengths)
    D ~ Truncated(Normal(100, 20), 0, Inf)
    σ ~ Truncated(Normal(1, 0.1), 0, Inf)
    β ~ Normal(0, 0.5)
    Turing.@addlogprob! composite_loglikelihood_power_density(D, β, σ, df, contig_lengths)
end
m = power_density(df, contig_lengths)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">DynamicPPL.Model{typeof(Main.power_density), (:df, :contig_lengths), (), (), Tuple{DataFrames.DataFrame, Vector{Float64}}, Tuple{}, DynamicPPL.DefaultContext}(Main.power_density, (df = 320×8 DataFrame
 Row │ DISTANCE  IBD_LEFT  IBD_RIGHT  IBD_MID  NR_PAIRS  COUNT  DISTANCE_INDEX ⋯
     │ Float64   Float64   Float64    Float64  Int64     Int64  Int64          ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │     10.0     0.04       0.041   0.0405         2      0               1 ⋯
   2 │     10.0     0.041      0.042   0.0415         2      0               1
   3 │     10.0     0.042      0.043   0.0425         2      0               1
   4 │     10.0     0.043      0.044   0.0435         2      0               1
   5 │     10.0     0.044      0.045   0.0445         2      0               1 ⋯
   6 │     10.0     0.045      0.046   0.0455         2      0               1
   7 │     10.0     0.046      0.047   0.0465         2      0               1
   8 │     10.0     0.047      0.048   0.0475         2      0               1
  ⋮  │    ⋮         ⋮          ⋮         ⋮        ⋮        ⋮          ⋮        ⋱
 314 │     20.0     0.193      0.194   0.1935         1      0               2 ⋯
 315 │     20.0     0.194      0.195   0.1945         1      0               2
 316 │     20.0     0.195      0.196   0.1955         1      0               2
 317 │     20.0     0.196      0.197   0.1965         1      0               2
 318 │     20.0     0.197      0.198   0.1975         1      0               2 ⋯
 319 │     20.0     0.198      0.199   0.1985         1      0               2
 320 │     20.0     0.199      0.2     0.1995         1      0               2
                                                   1 column and 305 rows omitted, contig_lengths = [1.0, 1.5]), NamedTuple(), DynamicPPL.DefaultContext())</code></pre><p>Sample 4 chains in a serial fashion.</p><pre><code class="language-julia hljs">chains = sample(m, NUTS(), MCMCSerial(), 1000, 4)
plot(chains)</code></pre><img src="d996d0ca.svg" alt="Example block output"/><p>More complex models can be built with cautious consideration of identifiability. For example, we can fit a piecewise exponential density model using a custom function and numerical integration.</p><pre><code class="language-julia hljs">function piecewise_D(t, parameters)
    De1, De2, alpha, t0 = parameters
    t &lt;= t0 ? De1 * exp(-alpha * t) : De2
end
@model function exponential_density(df, contig_lengths)
    De1 ~ Truncated(Normal(1000, 100), 0, Inf)
    De2 ~ Truncated(Normal(1000, 100), 0, Inf)
    α ~ Normal(0, 0.05)
    t0 ~ Truncated(Normal(500, 100), 0, Inf)
    σ ~ Truncated(Normal(1, 0.1), 0, Inf)
    theta = [De1, De2, α, t0]
    Turing.@addlogprob! composite_loglikelihood_custom(
        piecewise_D,
        theta,
        σ,
        df,
        contig_lengths,
    )
end
m = exponential_density(df, contig_lengths)
chain = sample(m, NUTS(), 1000;)
DataFrame(summarize(chain))</code></pre><div><div style = "float: left;"><span>5×8 DataFrame</span></div><div style = "clear: both;"></div></div><div class = "data-frame" style = "overflow-x: scroll;"><table class = "data-frame" style = "margin-bottom: 6px;"><thead><tr class = "columnLabelRow"><th class = "stubheadLabel" style = "font-weight: bold; text-align: right;">Row</th><th style = "text-align: left;">parameters</th><th style = "text-align: left;">mean</th><th style = "text-align: left;">std</th><th style = "text-align: left;">mcse</th><th style = "text-align: left;">ess_bulk</th><th style = "text-align: left;">ess_tail</th><th style = "text-align: left;">rhat</th><th style = "text-align: left;">ess_per_sec</th></tr><tr class = "columnLabelRow"><th class = "stubheadLabel" style = "font-weight: bold; text-align: right;"></th><th title = "Symbol" style = "text-align: left;">Symbol</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th></tr></thead><tbody><tr class = "dataRow"><td class = "rowLabel" style = "font-weight: bold; text-align: right;">1</td><td style = "text-align: left;">De1</td><td style = "text-align: right;">988.905</td><td style = "text-align: right;">97.9171</td><td style = "text-align: right;">4.5973</td><td style = "text-align: right;">455.942</td><td style = "text-align: right;">324.419</td><td style = "text-align: right;">1.01692</td><td style = "text-align: right;">1.05593</td></tr><tr class = "dataRow"><td class = "rowLabel" style = "font-weight: bold; text-align: right;">2</td><td style = "text-align: left;">De2</td><td style = "text-align: right;">997.941</td><td style = "text-align: right;">97.4677</td><td style = "text-align: right;">4.32365</td><td style = "text-align: right;">513.115</td><td style = "text-align: right;">343.077</td><td style = "text-align: right;">0.999456</td><td style = "text-align: right;">1.18834</td></tr><tr class = "dataRow"><td class = "rowLabel" style = "font-weight: bold; text-align: right;">3</td><td style = "text-align: left;">α</td><td style = "text-align: right;">0.0532756</td><td style = "text-align: right;">0.0232542</td><td style = "text-align: right;">0.00194388</td><td style = "text-align: right;">178.454</td><td style = "text-align: right;">282.08</td><td style = "text-align: right;">1.01164</td><td style = "text-align: right;">0.41329</td></tr><tr class = "dataRow"><td class = "rowLabel" style = "font-weight: bold; text-align: right;">4</td><td style = "text-align: left;">t0</td><td style = "text-align: right;">491.984</td><td style = "text-align: right;">97.6292</td><td style = "text-align: right;">4.63778</td><td style = "text-align: right;">435.649</td><td style = "text-align: right;">408.052</td><td style = "text-align: right;">0.999745</td><td style = "text-align: right;">1.00894</td></tr><tr class = "dataRow"><td class = "rowLabel" style = "font-weight: bold; text-align: right;">5</td><td style = "text-align: left;">σ</td><td style = "text-align: right;">1.0449</td><td style = "text-align: right;">0.0845457</td><td style = "text-align: right;">0.0036366</td><td style = "text-align: right;">523.785</td><td style = "text-align: right;">449.683</td><td style = "text-align: right;">0.999871</td><td style = "text-align: right;">1.21306</td></tr></tbody></table></div><hr/><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../overview/">« Theory overview</a><a class="docs-footer-nextpage" href="../inference/">Inference and model evaluation »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.16.1 on <span class="colophon-date" title="Wednesday 3 December 2025 08:08">Wednesday 3 December 2025</span>. Using Julia version 1.12.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
