<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Tutorial · IdentityByDescentDispersal.jl</title><meta name="title" content="Tutorial · IdentityByDescentDispersal.jl"/><meta property="og:title" content="Tutorial · IdentityByDescentDispersal.jl"/><meta property="twitter:title" content="Tutorial · IdentityByDescentDispersal.jl"/><meta name="description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="og:description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="twitter:description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="og:url" content="https://currocam.github.io/IdentityByDescentDispersal.jl/tutorial/"/><meta property="twitter:url" content="https://currocam.github.io/IdentityByDescentDispersal.jl/tutorial/"/><link rel="canonical" href="https://currocam.github.io/IdentityByDescentDispersal.jl/tutorial/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">IdentityByDescentDispersal.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>Tutorial</a><ul class="internal"><li><a class="tocitem" href="#Expected-number-of-identity-by-descent-blocks"><span>Expected number of identity-by-descent blocks</span></a></li><li><a class="tocitem" href="#Inference"><span>Inference</span></a></li></ul></li><li><span class="tocitem">Developer docs</span><ul><li><a class="tocitem" href="../95-reference/">Reference</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Tutorial</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Tutorial</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/main/docs/src/tutorial.jl#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Getting-started"><a class="docs-heading-anchor" href="#Getting-started">Getting started</a><a id="Getting-started-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-started" title="Permalink"></a></h1><p>This tutorial demonstrates basic usage of the package.</p><pre><code class="language-julia hljs">using IdentityByDescentDispersal</code></pre><h2 id="Expected-number-of-identity-by-descent-blocks"><a class="docs-heading-anchor" href="#Expected-number-of-identity-by-descent-blocks">Expected number of identity-by-descent blocks</a><a id="Expected-number-of-identity-by-descent-blocks-1"></a><a class="docs-heading-anchor-permalink" href="#Expected-number-of-identity-by-descent-blocks" title="Permalink"></a></h2><p>The main feature of this package is to compute the expected number of identity-by-descent blocks under different demographic scenarios. For example, we might explore how fast the number of blocks decay as a function of physical distance between pairs of individuals.</p><pre><code class="language-julia hljs">using Plots

L = 0.01         # Block length threshold (in Morgans)
G = 0.01         # Genome length (in Morgans)
D = 1.0          # Effective population density
σ = 1.0          # Dispersal rate
r_values = range(0.01, 25.0, length = 200);  # Distances

plot(
    r_values,
    expected_ibd_blocks_constant_density.(r_values, 1.0, 0.5, L, G),
    xlabel = &quot;Distance (r)&quot;,
    ylabel = &quot;E[IBD blocks]&quot;,
    title = &quot;Constant effective density scenario&quot;,
    label = &quot;D=1.0, σ=0.5&quot;,
)
plot!(
    r_values,
    expected_ibd_blocks_constant_density.(r_values, 2.0, 0.5, L, G),
    label = &quot;D=2.0, σ=0.5&quot;,
)
plot!(
    r_values,
    expected_ibd_blocks_constant_density.(r_values, 1.0, 0.8, L, G),
    label = &quot;D=1.0, σ=0.8&quot;,
)</code></pre><img src="a41dbf19.svg" alt="Example block output"/><p>In addition to constant effective densities and power density functions, we also support arbitrary effective density functions provided by the user via numerical integration. Next, we illustrate how to define a custom function by considering a scenario with an oscillating effective density function.</p><pre><code class="language-julia hljs">function De(t, θ)
    D₀, a, ω = θ
    D₀ * (1 + a * sin(ω * t))
end

θ = [1.0, 0.5, 2π]  # Parameters for De(t): D₀, a, ω
t_values = range(0.0, 10.0, length = 200)   # Time for plotting D_e(t)
ibd_values = [expected_ibd_blocks_custom(r, De, θ, σ, L, G) for r in r_values]
density_values = [De(t, θ) for t in t_values]
plot(
    plot(
        t_values,
        density_values,
        xlabel = &quot;Time (t)&quot;,
        ylabel = &quot;Effective Density Dₑ(t)&quot;,
        label = &quot;θ = [D₀=1, a=0.5, ω=2π]&quot;,
        title = &quot;Effective Density Trajectory&quot;,
    ),
    plot(
        r_values,
        ibd_values,
        xlabel = &quot;Distance (r)&quot;,
        ylabel = &quot;E[IBD blocks]&quot;,
        label = &quot;σ = 0.5, L = 0.01, G = 0.01&quot;,
        title = &quot;Expected Number of IBD Blocks&quot;,
    ),
    layout = (2, 1),
    size = (600, 800),
    title = &quot;User-defined effective density function&quot;,
)</code></pre><img src="4fab282e.svg" alt="Example block output"/><h2 id="Inference"><a class="docs-heading-anchor" href="#Inference">Inference</a><a id="Inference-1"></a><a class="docs-heading-anchor-permalink" href="#Inference" title="Permalink"></a></h2><p>Ringbauer proposed to do inference by assuming that the observed number of IBD blocks that a pair <span>$r$</span> units apart that fall within a small bin <span>$[L_i, L_i + \Delta L]$</span> follows a Poisson distribution with mean <span>$E[N_{L_i}(r, \theta)] \Delta L$</span>.</p><p>Therefore, the log probability of observing <code>Y</code> identity-by-descent blocks whose length fall in the bin <span>$[L_i, L_i + \Delta L]$</span> from a pair of individuals that are <span>$r$</span> units apart can be calculated simply as:</p><pre><code class="language-julia hljs">using Distributions
λ = expected_ibd_blocks_constant_density(
    0.2, # Distance,
    0.5, # Dispersal rate
    2.0, # Effective density
    0.01, # Block length threshold (in Morgans)
    1.0, # Genome length (in Morgans)
)
Y = 5 # Observed number of IBD blocks
logpdf(Poisson(λ), Y)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">-775.0267763054296</code></pre><p>The inference scheme is based on a composite-likelihood that treats each bin from different pairs of individuals as independent. For computational reasons, it is also faster to aggregate observations that correspond to the same distance and bin.</p><p>The input data for this sort of analysis is:</p><ul><li>A <code>DataFrame</code> containing the length of identity-by-descent blocks shared across different individuals</li><li>A <code>DataFrame</code> containing distances across pairs of individuals.</li><li>A list of contig lengths to properly take into account chromosomal edges (in Morgans).</li></ul><p>Preprocessing the data requires combining the different sources of information and binning the identity-by-descent blocks. For this purpose, we provide a helper function that returns a <code>DataFrame</code> in a &quot;long&quot; format.</p><pre><code class="language-julia hljs">using DataFrames
ibd_blocks = DataFrame(
    ID1 = [&quot;A&quot;, &quot;B&quot;, &quot;A&quot;, &quot;C&quot;, &quot;B&quot;],
    ID2 = [&quot;B&quot;, &quot;A&quot;, &quot;C&quot;, &quot;A&quot;, &quot;C&quot;],
    span = [0.005, 0.012, 0.21, 0.10, 0.08], # Morgans
)
individual_distances = DataFrame(
    ID1 = [&quot;A&quot;, &quot;A&quot;, &quot;B&quot;],
    ID2 = [&quot;B&quot;, &quot;C&quot;, &quot;C&quot;],
    distance = [10.0, 20.0, 10.0], # (e.g., in kilometers)
)
contig_lengths = [1.0, 1.5] # Contig lengths (in Morgans)
bins = [0.01, 0.02, 0.03] # Right edges of bins (in Morgans)
min_length = 0.005 # Minimum IBD length to consider
df = preprocess_dataset(ibd_blocks, individual_distances, bins, min_length)</code></pre><div><div style = "float: left;"><span>6×8 DataFrame</span></div><div style = "clear: both;"></div></div><div class = "data-frame" style = "overflow-x: scroll;"><table class = "data-frame" style = "margin-bottom: 6px;"><thead><tr class = "header"><th class = "rowNumber" style = "font-weight: bold; text-align: right;">Row</th><th style = "text-align: left;">DISTANCE</th><th style = "text-align: left;">IBD_LEFT</th><th style = "text-align: left;">IBD_RIGHT</th><th style = "text-align: left;">IBD_MID</th><th style = "text-align: left;">NR_PAIRS</th><th style = "text-align: left;">COUNT</th><th style = "text-align: left;">DISTANCE_INDEX</th><th style = "text-align: left;">BIN_INDEX</th></tr><tr class = "subheader headerLastRow"><th class = "rowNumber" style = "font-weight: bold; text-align: right;"></th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Int64" style = "text-align: left;">Int64</th><th title = "Int64" style = "text-align: left;">Int64</th><th title = "Int64" style = "text-align: left;">Int64</th><th title = "Int64" style = "text-align: left;">Int64</th></tr></thead><tbody><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">1</td><td style = "text-align: right;">10.0</td><td style = "text-align: right;">0.005</td><td style = "text-align: right;">0.01</td><td style = "text-align: right;">0.0075</td><td style = "text-align: right;">2</td><td style = "text-align: right;">1</td><td style = "text-align: right;">1</td><td style = "text-align: right;">1</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">2</td><td style = "text-align: right;">10.0</td><td style = "text-align: right;">0.01</td><td style = "text-align: right;">0.02</td><td style = "text-align: right;">0.015</td><td style = "text-align: right;">2</td><td style = "text-align: right;">1</td><td style = "text-align: right;">1</td><td style = "text-align: right;">2</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">3</td><td style = "text-align: right;">10.0</td><td style = "text-align: right;">0.02</td><td style = "text-align: right;">0.03</td><td style = "text-align: right;">0.025</td><td style = "text-align: right;">2</td><td style = "text-align: right;">0</td><td style = "text-align: right;">1</td><td style = "text-align: right;">3</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">4</td><td style = "text-align: right;">20.0</td><td style = "text-align: right;">0.005</td><td style = "text-align: right;">0.01</td><td style = "text-align: right;">0.0075</td><td style = "text-align: right;">1</td><td style = "text-align: right;">0</td><td style = "text-align: right;">2</td><td style = "text-align: right;">1</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">5</td><td style = "text-align: right;">20.0</td><td style = "text-align: right;">0.01</td><td style = "text-align: right;">0.02</td><td style = "text-align: right;">0.015</td><td style = "text-align: right;">1</td><td style = "text-align: right;">0</td><td style = "text-align: right;">2</td><td style = "text-align: right;">2</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">6</td><td style = "text-align: right;">20.0</td><td style = "text-align: right;">0.02</td><td style = "text-align: right;">0.03</td><td style = "text-align: right;">0.025</td><td style = "text-align: right;">1</td><td style = "text-align: right;">0</td><td style = "text-align: right;">2</td><td style = "text-align: right;">3</td></tr></tbody></table></div><p>We can now use the df <code>DataFrame</code>to compute composite loglikelihoods of different parameters using the family of <code>composite_loglikelihood_*</code> functions. The second major feature of this package is that functions are all compatible with automatic differentiation. Analytical solutions depend on a modified Bessel function of the second kind, which is not widely implemented in a way that allows for automatic differentiation . Here, we rely on the <code>BesselK.jl</code> package for doing this. In addition, we rely on the system of <code>QuadGK.jl</code> for numerical integration compatible with automatic differentiation. This allows us to perform gradient-based optimization and interact with existing software.</p><p>We recommend using this package together with the <code>Turing.jl</code>, as it provides a convenient interface for Bayesian and frequentist inference.</p><h3 id="Maximum-likelihood-estimation"><a class="docs-heading-anchor" href="#Maximum-likelihood-estimation">Maximum likelihood estimation</a><a id="Maximum-likelihood-estimation-1"></a><a class="docs-heading-anchor-permalink" href="#Maximum-likelihood-estimation" title="Permalink"></a></h3><p>For example, we can estimate the parameters of a constant density model using maximum likelihood estimation:</p><pre><code class="language-julia hljs">using Turing, StatsBase, StatsPlots
@model function constant_density(df, contig_lengths)
    D ~ Uniform(0, 100)
    σ ~ Uniform(0, 20)
    Turing.@addlogprob! composite_loglikelihood_constant_density(D, σ, df, contig_lengths)
end</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">constant_density (generic function with 2 methods)</code></pre><p>Generate a MLE estimate.</p><pre><code class="language-julia hljs">mle_estimate = maximum_likelihood(constant_density(df, contig_lengths))
DataFrame(coeftable(mle_estimate)) # computed from the Fisher information matrix</code></pre><div><div style = "float: left;"><span>2×7 DataFrame</span></div><div style = "clear: both;"></div></div><div class = "data-frame" style = "overflow-x: scroll;"><table class = "data-frame" style = "margin-bottom: 6px;"><thead><tr class = "header"><th class = "rowNumber" style = "font-weight: bold; text-align: right;">Row</th><th style = "text-align: left;">Name</th><th style = "text-align: left;">Coef.</th><th style = "text-align: left;">Std. Error</th><th style = "text-align: left;">z</th><th style = "text-align: left;">Pr(&gt;|z|)</th><th style = "text-align: left;">Lower 95%</th><th style = "text-align: left;">Upper 95%</th></tr><tr class = "subheader headerLastRow"><th class = "rowNumber" style = "font-weight: bold; text-align: right;"></th><th title = "String" style = "text-align: left;">String</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th></tr></thead><tbody><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">1</td><td style = "text-align: left;">D</td><td style = "text-align: right;">100.0</td><td style = "text-align: right;">375.837</td><td style = "text-align: right;">0.266073</td><td style = "text-align: right;">0.790183</td><td style = "text-align: right;">-636.627</td><td style = "text-align: right;">836.627</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">2</td><td style = "text-align: left;">σ</td><td style = "text-align: right;">1.08615</td><td style = "text-align: right;">3.15893</td><td style = "text-align: right;">0.343834</td><td style = "text-align: right;">0.730971</td><td style = "text-align: right;">-5.10525</td><td style = "text-align: right;">7.27755</td></tr></tbody></table></div><h3 id="Bayesian-inference"><a class="docs-heading-anchor" href="#Bayesian-inference">Bayesian inference</a><a id="Bayesian-inference-1"></a><a class="docs-heading-anchor-permalink" href="#Bayesian-inference" title="Permalink"></a></h3><p>Alternatively, we can do a standard Bayesian inference with any of the available inference algorithms such as NUTS. Here, we fit a power-density model using NUTS:</p><pre><code class="language-julia hljs">@model function power_density(df, contig_lengths)
    D ~ Truncated(Normal(100, 20), 0, Inf)
    σ ~ Truncated(Normal(1, 0.1), 0, Inf)
    β ~ Normal(0, 0.5)
    Turing.@addlogprob! composite_loglikelihood_power_density(D, β, σ, df, contig_lengths)
end
m = power_density(df, contig_lengths)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">DynamicPPL.Model{typeof(Main.power_density), (:df, :contig_lengths), (), (), Tuple{DataFrames.DataFrame, Vector{Float64}}, Tuple{}, DynamicPPL.DefaultContext}(Main.power_density, (df = <span class="sgr1">6×8 DataFrame
 Row │ DISTANCE  IBD_LEFT  IBD_RIGHT  IBD_MID  NR_PAIRS  COUNT  DISTANCE_INDEX ⋯
     │<span class="sgr90"> Float64   Float64   Float64    Float64  Int64     Int64  Int64          ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │     10.0     0.005       0.01   0.0075         2      1               1 ⋯
   2 │     10.0     0.01        0.02   0.015          2      1               1
   3 │     10.0     0.02        0.03   0.025          2      0               1
   4 │     20.0     0.005       0.01   0.0075         1      0               2
   5 │     20.0     0.01        0.02   0.015          1      0               2 ⋯
   6 │     20.0     0.02        0.03   0.025          1      0               2
</span><span class="sgr36">                                                                1 column omitted, contig_lengths = [1.0, 1.5]), NamedTuple(), DynamicPPL.DefaultContext())</span></span></code></pre><p>Sample 4 chains in a serial fashion.</p><pre><code class="language-julia hljs">chains = sample(m, NUTS(), MCMCSerial(), 1000, 4)
plot(chains)</code></pre><img src="3d4f65c3.svg" alt="Example block output"/><p>More complex models can be built with cautious consideration of identifiability. For example, we can fit a piecewise exponential density model using a custom function and numerical integration.</p><pre><code class="language-julia hljs">function piecewise_D(t, parameters)
    De1, De2, alpha, t0 = parameters
    t &lt;= t0 ? De1*exp(-alpha*t) : De2
end
@model function exponential_density(df, contig_lengths)
    De1 ~ Truncated(Normal(1000, 100), 0, Inf)
    De2 ~ Truncated(Normal(1000, 100), 0, Inf)
    α ~ Normal(0, 0.05)
    t0 ~ Truncated(Normal(500, 100), 0, Inf)
    σ ~ Truncated(Normal(1, 0.1), 0, Inf)
    theta = [De1, De2, α, t0]
    Turing.@addlogprob! composite_loglikelihood_custom(
        piecewise_D,
        theta,
        σ,
        df,
        contig_lengths,
    )
end
m = exponential_density(df, contig_lengths)
chain = sample(m, NUTS(), 1000;)
DataFrame(summarize(chain))</code></pre><div><div style = "float: left;"><span>5×8 DataFrame</span></div><div style = "clear: both;"></div></div><div class = "data-frame" style = "overflow-x: scroll;"><table class = "data-frame" style = "margin-bottom: 6px;"><thead><tr class = "header"><th class = "rowNumber" style = "font-weight: bold; text-align: right;">Row</th><th style = "text-align: left;">parameters</th><th style = "text-align: left;">mean</th><th style = "text-align: left;">std</th><th style = "text-align: left;">mcse</th><th style = "text-align: left;">ess_bulk</th><th style = "text-align: left;">ess_tail</th><th style = "text-align: left;">rhat</th><th style = "text-align: left;">ess_per_sec</th></tr><tr class = "subheader headerLastRow"><th class = "rowNumber" style = "font-weight: bold; text-align: right;"></th><th title = "Symbol" style = "text-align: left;">Symbol</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th></tr></thead><tbody><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">1</td><td style = "text-align: left;">De1</td><td style = "text-align: right;">996.353</td><td style = "text-align: right;">102.844</td><td style = "text-align: right;">4.6557</td><td style = "text-align: right;">492.222</td><td style = "text-align: right;">510.248</td><td style = "text-align: right;">0.99988</td><td style = "text-align: right;">14.7611</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">2</td><td style = "text-align: left;">De2</td><td style = "text-align: right;">1001.79</td><td style = "text-align: right;">99.5832</td><td style = "text-align: right;">3.85642</td><td style = "text-align: right;">663.446</td><td style = "text-align: right;">577.821</td><td style = "text-align: right;">0.999222</td><td style = "text-align: right;">19.8958</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">3</td><td style = "text-align: left;">α</td><td style = "text-align: right;">0.00644892</td><td style = "text-align: right;">0.00500065</td><td style = "text-align: right;">0.000267806</td><td style = "text-align: right;">413.072</td><td style = "text-align: right;">385.658</td><td style = "text-align: right;">1.00877</td><td style = "text-align: right;">12.3875</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">4</td><td style = "text-align: left;">t0</td><td style = "text-align: right;">474.198</td><td style = "text-align: right;">108.566</td><td style = "text-align: right;">5.69232</td><td style = "text-align: right;">365.041</td><td style = "text-align: right;">225.768</td><td style = "text-align: right;">1.00639</td><td style = "text-align: right;">10.9471</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">5</td><td style = "text-align: left;">σ</td><td style = "text-align: right;">0.98481</td><td style = "text-align: right;">0.101815</td><td style = "text-align: right;">0.0046525</td><td style = "text-align: right;">483.788</td><td style = "text-align: right;">332.633</td><td style = "text-align: right;">1.00283</td><td style = "text-align: right;">14.5081</td></tr></tbody></table></div><hr/><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../95-reference/">Reference »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.13.0 on <span class="colophon-date" title="Tuesday 24 June 2025 08:28">Tuesday 24 June 2025</span>. Using Julia version 1.11.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
