<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Theory overview · IdentityByDescentDispersal.jl</title><meta name="title" content="Theory overview · IdentityByDescentDispersal.jl"/><meta property="og:title" content="Theory overview · IdentityByDescentDispersal.jl"/><meta property="twitter:title" content="Theory overview · IdentityByDescentDispersal.jl"/><meta name="description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="og:description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="twitter:description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="og:url" content="https://currocam.github.io/IdentityByDescentDispersal.jl/overview/"/><meta property="twitter:url" content="https://currocam.github.io/IdentityByDescentDispersal.jl/overview/"/><link rel="canonical" href="https://currocam.github.io/IdentityByDescentDispersal.jl/overview/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">IdentityByDescentDispersal.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>Theory overview</a><ul class="internal"><li><a class="tocitem" href="#Diffusion-approximation-of-dispersal"><span>Diffusion approximation of dispersal</span></a></li><li><a class="tocitem" href="#Coalescent-theory"><span>Coalescent theory</span></a></li><li><a class="tocitem" href="#IBD-blocks"><span>IBD blocks</span></a></li><li><a class="tocitem" href="#Inference-scheme"><span>Inference scheme</span></a></li></ul></li><li><a class="tocitem" href="../tutorial/">Basic usage</a></li><li><a class="tocitem" href="../inference/">Inference and model evaluation</a></li><li><span class="tocitem">Developer docs</span><ul><li><a class="tocitem" href="../95-reference/">Reference</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Theory overview</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Theory overview</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/main/docs/src/overview.jl#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Theory-overview"><a class="docs-heading-anchor" href="#Theory-overview">Theory overview</a><a id="Theory-overview-1"></a><a class="docs-heading-anchor-permalink" href="#Theory-overview" title="Permalink"></a></h1><p>This package provides an efficient implementation of the inference scheme proposed by Ringbauer et al. ( 2017) to estimate the mean dispersal rate and the effective population density of a population.  Here, we present an overview of the theory behind the method, but we refer to the original publication for details.</p><p>Mates tend to live close to each other and their offspring. This results in an inverse correlation between geographical distance and genetic relatedness. The goal of this package is to do demographic inference based on such spatial genetic patterns.</p><h2 id="Diffusion-approximation-of-dispersal"><a class="docs-heading-anchor" href="#Diffusion-approximation-of-dispersal">Diffusion approximation of dispersal</a><a id="Diffusion-approximation-of-dispersal-1"></a><a class="docs-heading-anchor-permalink" href="#Diffusion-approximation-of-dispersal" title="Permalink"></a></h2><p>First, let’s provide some intuition on how we model the dispersal of individuals in continuous space.  From an ecological perspective, we might be interested in modeling the details of single-generation dispersal, such as the mating process or differences in dispersal across individuals and sexes.  Such details are, however, not generally feasible to estimate from genetic data. Instead, we aim to estimate the <strong>mean dispersal rate</strong>, this is for a given gene in a given individual the expected euclidean distance to the parent from which the gene was inherited.</p><p>The method implemented here approximates the spatial movement of genetic material using a diffusion process. Let <span>$r(t)$</span> be the distance between two homologous gene copies at a time <span>$t$</span> before the present. First, note that when <span>$t=\text{TMRCA}$</span> (time to the most common recent ancestor), then <span>$r(\text{TMRCA}) = 0$</span>. The physical distance between both homologs at the present, <span>$r(0)=r_{\text{obs}}$</span> (i.e. the distance we can observe), is the sum of the sequence of displacements due to dispersal since the time to the most common recent ancestor.</p><p>If dispersal distances are independent and identically distributed, then, by the central-limit theorem, the random variable <span>$r_{\text{obs}}$</span> converges to a Gaussian distribution with axial variance of <span>$\sigma^2\times \text{TMRCA}$</span> as <span>$\text{TMRCA}$</span> grows. Here <span>$\sigma^2$</span> is the average squared axial parent-offspring distance (i.e. the dispersal rate).</p><p>This result is independent of the details of the single-generation dispersal process, as long as the variance of the dispersal kernel is finite. For example, let’s consider a population where dispersal across a single-generation dispersal can be modeled using a uniform distribution. Below, we visualize the distribution of distances between homologous loci when <span>$t = 1$</span> (single-generation dispersal) and <span>$t=30$</span>.  For simplicity, we assume 1-dimensional space. As we can see, the approximation becomes very accurate even after a few generations.</p><pre><code class="language-julia hljs">using Distributions, Plots, StatsPlots
displacement = Uniform(-1, 1)
σ2 = var(displacement) # Theoretical variance of displacement
distance(t) = sum(rand(displacement, t)); # Distance after t migrations
n_draws = 10_000
p1 = histogram(
    [distance(1) for _ = 1:n_draws],
    normalize = true,
    label = &quot;t = 1&quot;,
    xlabel = &quot;Distance&quot;,
    ylabel = &quot;Density&quot;,
)
plot!(p1, Normal(0, sqrt(σ2 * 1)), lw = 2, label = &quot;Gaussian Approximation&quot;)

p2 = histogram(
    [distance(30) for _ = 1:n_draws],
    normalize = true,
    label = &quot;t = 30&quot;,
    xlabel = &quot;Distance&quot;,
    ylabel = &quot;Density&quot;,
)
plot!(p2, Normal(0, sqrt(σ2 * 30)), lw = 2, label = &quot;Gaussian Approximation&quot;)
plot(p1, p2, layout = (1, 2), size = (900, 400), legend = :topright, framestyle = :box)</code></pre><img src="18a77227.svg" alt="Example block output"/><p>This approximation is expected to work well in regimes where the time to the most common recent ancestor is not extremely small and when dispersal is not correlated across generations and is time-homogeneous (i.e. displacements are independent and identically distributed).</p><p>From the previous section, it is clear that it would be possible to design an inference scheme based on the diffusion approximation if we knew the exact time to the most recent common ancestor between homologs. For example, the probability of two homologs being more than one unit apart given that they coalesced 20 generations ago can be calculated from the density of the Gaussian approximation for any <span>$\sigma$</span>.</p><pre><code class="language-julia hljs">1 - cdf(Normal(0, sqrt(σ2 * 20)), 1)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">0.34926767915166934</code></pre><h2 id="Coalescent-theory"><a class="docs-heading-anchor" href="#Coalescent-theory">Coalescent theory</a><a id="Coalescent-theory-1"></a><a class="docs-heading-anchor-permalink" href="#Coalescent-theory" title="Permalink"></a></h2><p>However, the time to the most common recent ancestor is not directly observable. Instead, what we can do is marginalize across all possible times <span>$[0, \infty)$</span>. Here one can use standard coalescent theory to model the time to coalescence. Roughly speaking, we consider a scenario where two homologous loci <em>might</em> coalesce when they become very close. The rate with which they coalesce when they are very close is inversely proportional to local effective density <span>$D_e$</span> (effective number of individuals per unit of space).  According to this model, the coalescence rate at time <span>$t$</span> before the present for two homologs that are currently a distance <span>$d$</span> apart is <span>$p(r(t)=0|r(0)=d) \times \frac{1}{D_e(t)}$</span>, i.e. the rate at which they come close enough times the rate at which they coalesce given that they are close enough.</p><pre><code class="language-julia hljs">generations = 1:300
D_e = 1 # Constant effective density of 1 (individuals per unit area)
σ = 1 # Constant dispersal rate of 1( distance units per generation)
r = 3 # Pairwise distance between loci
densities = [pdf(Normal(0, sqrt(σ^2 * t)), r) / D_e for t in generations]
plot(
    generations,
    densities,
    ylabel = &quot;Coalescence rate ϕ(t)&quot;,
    xlabel = &quot;Generations&quot;,
    label = &quot;(D=1, σ=1, r=3)&quot;,
)</code></pre><img src="f221e597.svg" alt="Example block output"/><h2 id="IBD-blocks"><a class="docs-heading-anchor" href="#IBD-blocks">IBD blocks</a><a id="IBD-blocks-1"></a><a class="docs-heading-anchor-permalink" href="#IBD-blocks" title="Permalink"></a></h2><p>The last ingredient of this method is the relationship between identity-by-descent (IBD) blocks and time to the most recent common ancestor. First, we point out that generally only recent coalescent events are informative of the dispersal rate. This is the reason why this inference scheme relies on identity-by-descent blocks, as such data are more informative about the recent past than polymorphism data.</p><p>We say that a segment of DNA is a shared identity-by-descent block if it has been inherited from a common ancestor without being broken up by recombination. A shared identity-by-descent block of length <span>$l$</span> that finds its common ancestor at time <span>$t$</span> has experienced <span>$2t$</span> meiosis. Therefore, we expect <span>$l$</span> to decay with <span>$t$</span>.</p><p>If we model recombination as a Poisson process and measure genetic distance in Morgans (as is typically done), then the probability that a region of length <span>$L$</span> does not recombine can be obtained from the exponential distribution with rate <span>$\exp(-2Lt)$</span></p><p>Ringbauer et al. (2017) derived that the expected density of identity-by-descent blocks of length <span>$l$</span> per pair of haploid individuals and per unit of block length given that the time to the most recent common ancestor is <span>$t$</span> is given by</p><p class="math-container">\[E[K_l | t] \approx G 4 t^2 \exp(-2lt)\]</p><p>where <span>$G$</span> is the length of the genome in Morgans. This equation is obtained by accounting for the probability that there is a region of length <span>$l$</span> that has not recombined (the <span>$\exp(-2lt)$</span> term) which is delimited by two recombination events (the <span>$4 t^2$</span> term) and summing across all possible start sites (the <span>$G$</span> factor). A slightly more complex expression that accounts for the effect of chromosomal edges and diploidy is provided in the article and the software implementation.</p><h2 id="Inference-scheme"><a class="docs-heading-anchor" href="#Inference-scheme">Inference scheme</a><a id="Inference-scheme-1"></a><a class="docs-heading-anchor-permalink" href="#Inference-scheme" title="Permalink"></a></h2><p>As mentioned above, the inference scheme relies on computing the expected density of identity-by-descent blocks under a given demographic model. This involves marginalizing across all possible times of coalescence, therefore computing integrals in the form of</p><p class="math-container">\[\mathbb{E}[L] = \int_0^\infty \mathbb{E}[L | t] \phi(t) dt\]</p><p>where <span>$\phi(t)$</span> is the probability density function of the time to the most recent common ancestor. This expression is hard to track down analytically. Ringbauer et al. (2017) derived an analytical solution for a family of demographic models. Alternatively, this software implementation provides a numerical approximation of the integral using Gaussian-quadrature rules. We refer to the other sections of the documentation for more details.</p><p>Ringbauer et al. (2017) proposed an inference scheme where they assumed the number of observed shared identity-by-descent blocks whose length fall in a small bin follows a Poisson distribution. The rate of the distribution can be calculated from the expected density of identity-by-descent blocks under the demographic model of interest. If the bin is small enough, it can be calculated as</p><p class="math-container">\[\mathbb{E}[\#\{L, L+\Delta L\}] = \mathbb{E}[L] \Delta L\]</p><p>Finally, they propose to approximate the likelihood of the observed data across many bins and pairs of individuals using a composite likelihood (that is, by assuming pairwise observations are independent).</p><p class="math-container">\[L(\theta | Y_i^j) = \Pr(K = Y_i^j | \mathbb{E}[L_i]^j \Delta L)\]</p><p>where <span>$Y_i^j$</span> is the number of observed shared-identity-by-descent blocks whose length fall in the <span>$i$</span>th bin and are shared by the <span>$j$</span>th pair of individuals. They then define the composite likelihood as</p><p class="math-container">\[CL(\theta | \text{Data}) = \prod_{i,j} \Pr(K = Y_i^j | \mathbb{E}[L_i]^j \Delta L)\]</p><hr/><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../tutorial/">Basic usage »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Thursday 9 October 2025 16:34">Thursday 9 October 2025</span>. Using Julia version 1.12.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
