<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API reference · IdentityByDescentDispersal.jl</title><meta name="title" content="API reference · IdentityByDescentDispersal.jl"/><meta property="og:title" content="API reference · IdentityByDescentDispersal.jl"/><meta property="twitter:title" content="API reference · IdentityByDescentDispersal.jl"/><meta name="description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="og:description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="twitter:description" content="Documentation for IdentityByDescentDispersal.jl."/><meta property="og:url" content="https://currocam.github.io/IdentityByDescentDispersal.jl/95-reference/"/><meta property="twitter:url" content="https://currocam.github.io/IdentityByDescentDispersal.jl/95-reference/"/><link rel="canonical" href="https://currocam.github.io/IdentityByDescentDispersal.jl/95-reference/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">IdentityByDescentDispersal.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../overview/">Theory overview</a></li><li><a class="tocitem" href="../tutorial/">Basic usage</a></li><li><a class="tocitem" href="../inference/">Inference and model evaluation</a></li><li><span class="tocitem">Developer docs</span><ul><li class="is-active"><a class="tocitem" href>API reference</a><ul class="internal"><li><a class="tocitem" href="#Contents"><span>Contents</span></a></li><li><a class="tocitem" href="#Index"><span>Index</span></a></li></ul></li></ul></li><li><a class="tocitem" href="../references/">References</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Developer docs</a></li><li class="is-active"><a href>API reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API reference</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/main/docs/src/95-reference.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="API-reference"><a class="docs-heading-anchor" href="#API-reference">API reference</a><a id="API-reference-1"></a><a class="docs-heading-anchor-permalink" href="#API-reference" title="Permalink"></a></h1><h2 id="Contents"><a class="docs-heading-anchor" href="#Contents">Contents</a><a id="Contents-1"></a><a class="docs-heading-anchor-permalink" href="#Contents" title="Permalink"></a></h2><ul><li><a href="#API-reference">API reference</a></li><li class="no-marker"><ul><li><a href="#Contents">Contents</a></li><li><a href="#Index">Index</a></li></ul></li></ul><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#IdentityByDescentDispersal.age_density_ibd_blocks_custom"><code>IdentityByDescentDispersal.age_density_ibd_blocks_custom</code></a></li><li><a href="#IdentityByDescentDispersal.composite_loglikelihood_constant_density-Tuple{Real, Real, DataFrames.DataFrame, AbstractArray{&lt;:Real}}"><code>IdentityByDescentDispersal.composite_loglikelihood_constant_density</code></a></li><li><a href="#IdentityByDescentDispersal.composite_loglikelihood_custom-Tuple{Function, AbstractArray, Real, DataFrames.DataFrame, AbstractArray{&lt;:Real}}"><code>IdentityByDescentDispersal.composite_loglikelihood_custom</code></a></li><li><a href="#IdentityByDescentDispersal.composite_loglikelihood_power_density-Tuple{Real, Real, Real, DataFrames.DataFrame, AbstractArray{&lt;:Real}}"><code>IdentityByDescentDispersal.composite_loglikelihood_power_density</code></a></li><li><a href="#IdentityByDescentDispersal.default_ibd_bins-Tuple{}"><code>IdentityByDescentDispersal.default_ibd_bins</code></a></li><li><a href="#IdentityByDescentDispersal.expected_ibd_blocks_constant_density"><code>IdentityByDescentDispersal.expected_ibd_blocks_constant_density</code></a></li><li><a href="#IdentityByDescentDispersal.expected_ibd_blocks_custom"><code>IdentityByDescentDispersal.expected_ibd_blocks_custom</code></a></li><li><a href="#IdentityByDescentDispersal.expected_ibd_blocks_power_density"><code>IdentityByDescentDispersal.expected_ibd_blocks_power_density</code></a></li><li><a href="#IdentityByDescentDispersal.preprocess_dataset-Tuple{DataFrames.DataFrame, DataFrames.DataFrame, AbstractVector, Real}"><code>IdentityByDescentDispersal.preprocess_dataset</code></a></li><li><a href="#IdentityByDescentDispersal.probability_coalescence-Tuple{Real, Real, Function, Real}"><code>IdentityByDescentDispersal.probability_coalescence</code></a></li><li><a href="#IdentityByDescentDispersal.safe_adbesselk-Tuple{Any, Any}"><code>IdentityByDescentDispersal.safe_adbesselk</code></a></li></ul><article><details class="docstring" open="true"><summary id="IdentityByDescentDispersal.age_density_ibd_blocks_custom"><a class="docstring-binding" href="#IdentityByDescentDispersal.age_density_ibd_blocks_custom"><code>IdentityByDescentDispersal.age_density_ibd_blocks_custom</code></a> — <span class="docstring-category">Function</span></summary><section><div><pre><code class="language-julia hljs">age_density_ibd_blocks_custom(t::Real, r::Real, De::Function, parameters::AbstractArray, sigma::Real, L::Real, G::Real, chromosomal_edges::Bool = true, diploid::Bool = true)</code></pre><p>Computes the expected density of identity-by-descent (IBD) blocks of length <code>L</code> and age <code>t</code> for a model where the effective population density is given by a custom function <code>De(t, parameters)</code>. This function returns the expected number of IBD blocks of age <code>t</code> per pair of individuals and per unit of block length.</p><ul><li><code>t</code> is the age of the IBD block,</li><li><code>r</code> is the geographic distance between samples,</li><li><code>De</code> is a user-defined function that takes time <code>t</code> and a <code>parameters</code> and returns the effective population density at time <code>t</code>.</li><li><code>parameters</code> is a user-defined array of parameters that the function <code>De</code> depends on.</li><li><code>sigma</code> is the root mean square dispersal distance per generation,</li><li><code>L</code> is the length of the IBD block (in Morgans),</li><li><code>G</code> is the total map length of the genome (in Morgans),</li></ul><p>If <code>chromosomal_edges</code> is <code>true</code> (the default), we account for chromosomal edge effects. If <code>diploid</code> is <code>true</code>, we multiply by a factor of 4 to account for the fact that each individual has two copies of each chromosome.</p><p>We calculate it as</p><p class="math-container">\[\mathbb{E}[N_L^t] =  G  4t^2 \exp(-2Lt) \cdot \Phi(t)\]</p><p>where <span>$\phi(t)$</span> is the probability that two homologous loci coalesce <span>$t$</span> generations ago.</p><p>Reference: Ringbauer, H., Coop, G., &amp; Barton, N. H. (2017). Genetics, 205(3), 1335–1351.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/ff93e5c6f8ea6a815166df0c3cf608d0d74cddd4/src/IdentityByDescentDispersal.jl#L673-L698">source</a></section></details></article><article><details class="docstring" open="true"><summary id="IdentityByDescentDispersal.composite_loglikelihood_constant_density-Tuple{Real, Real, DataFrames.DataFrame, AbstractArray{&lt;:Real}}"><a class="docstring-binding" href="#IdentityByDescentDispersal.composite_loglikelihood_constant_density-Tuple{Real, Real, DataFrames.DataFrame, AbstractArray{&lt;:Real}}"><code>IdentityByDescentDispersal.composite_loglikelihood_constant_density</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">composite_loglikelihood_constant_density(D::Real, sigma::Real, df::DataFrame, contig_lengths::AbstractArray{&lt;:Real}, chromosomal_edges::Bool=true, diploid::Bool=true) -&gt; Real</code></pre><p>Computes the composite log-likelihood of the observed IBD blocks under a model with constant population density.</p><ul><li><code>D</code>: Effective population density (diploid individuals per unit area).</li><li><code>sigma</code>: Root mean square dispersal distance per generation.</li><li><code>df</code>: DataFrame containing the observed IBD blocks in the format returned by <code>preprocess_dataset</code>.</li><li><code>contig_lengths</code>: Array of contig lengths in Morgans.</li></ul><p>Optionally:</p><ul><li><code>chromosomal_edges</code>: Whether to account for chromosomal edge effects.</li><li><code>diploid</code>: Whether to account for diploidy.</li><li><code>verbose</code>: Whether to print warnings when the computation fails.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/ff93e5c6f8ea6a815166df0c3cf608d0d74cddd4/src/IdentityByDescentDispersal.jl#L397-L411">source</a></section></details></article><article><details class="docstring" open="true"><summary id="IdentityByDescentDispersal.composite_loglikelihood_custom-Tuple{Function, AbstractArray, Real, DataFrames.DataFrame, AbstractArray{&lt;:Real}}"><a class="docstring-binding" href="#IdentityByDescentDispersal.composite_loglikelihood_custom-Tuple{Function, AbstractArray, Real, DataFrames.DataFrame, AbstractArray{&lt;:Real}}"><code>IdentityByDescentDispersal.composite_loglikelihood_custom</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">composite_loglikelihood_custom(De::Function, parameters::AbstractArray, sigma::Real, df::DataFrame, contig_lengths::AbstractArray{&lt;:Real}, chromosomal_edges::Bool=true, diploid::Bool=true) -&gt; Real</code></pre><p>Computes the composite log-likelihood of the observed IBD blocks under a model with constant population density.</p><ul><li><code>De</code> is  a user-defined function that takes time <code>t</code> and <code>parameters</code> and returns the effective population density at time <code>t</code>.</li><li><code>parameters</code> is a user-defined array of parameters that the function <code>De</code> depends on.</li><li><code>sigma</code>: Root mean square dispersal distance per generation.</li><li><code>df</code>: DataFrame containing the observed IBD blocks in the format returned by <code>preprocess_dataset</code>.</li><li><code>contig_lengths</code>: Array of contig lengths in Morgans.</li></ul><p>Optionally:</p><ul><li><code>chromosomal_edges</code>: Whether to account for chromosomal edge effects.</li><li><code>diploid</code>: Whether to account for diploidy.</li><li><code>verbose</code>: Whether to print warnings when the computation fails.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/ff93e5c6f8ea6a815166df0c3cf608d0d74cddd4/src/IdentityByDescentDispersal.jl#L575-L590">source</a></section></details></article><article><details class="docstring" open="true"><summary id="IdentityByDescentDispersal.composite_loglikelihood_power_density-Tuple{Real, Real, Real, DataFrames.DataFrame, AbstractArray{&lt;:Real}}"><a class="docstring-binding" href="#IdentityByDescentDispersal.composite_loglikelihood_power_density-Tuple{Real, Real, Real, DataFrames.DataFrame, AbstractArray{&lt;:Real}}"><code>IdentityByDescentDispersal.composite_loglikelihood_power_density</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">composite_loglikelihood_power_density(D::Real, beta::Real, sigma::Real, df::DataFrame, contig_lengths::AbstractArray{&lt;:Real}, chromosomal_edges::Bool=true, diploid::Bool=true) -&gt; Real</code></pre><p>Computes the composite log-likelihood of the observed IBD blocks under a model with constant population density.</p><ul><li><code>D</code>: Effective population density (diploid individuals per unit area).</li><li><code>beta</code> is the power of the density function,</li><li><code>sigma</code>: Root mean square dispersal distance per generation.</li><li><code>df</code>: DataFrame containing the observed IBD blocks in the format returned by <code>preprocess_dataset</code>.</li><li><code>contig_lengths</code>: Array of contig lengths in Morgans.</li></ul><p>Optionally:</p><ul><li><code>chromosomal_edges</code>: Whether to account for chromosomal edge effects.</li><li><code>diploid</code>: Whether to account for diploidy.</li><li><code>verbose</code>: Whether to print warnings when the computation fails.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/ff93e5c6f8ea6a815166df0c3cf608d0d74cddd4/src/IdentityByDescentDispersal.jl#L483-L498">source</a></section></details></article><article><details class="docstring" open="true"><summary id="IdentityByDescentDispersal.default_ibd_bins-Tuple{}"><a class="docstring-binding" href="#IdentityByDescentDispersal.default_ibd_bins-Tuple{}"><code>IdentityByDescentDispersal.default_ibd_bins</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">default_ibd_bins()</code></pre><p>Returns the default bins used in Ringbauer et. al. for identity-by-descent (IBD) block analysis, as well as the minimum length of IBD blocks to consider in Morgans.</p><p>Ringbauer, H., Coop, G. &amp; Barton, N.H. Genetics 205, 1335–1351 (2017).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/ff93e5c6f8ea6a815166df0c3cf608d0d74cddd4/src/IdentityByDescentDispersal.jl#L295-L300">source</a></section></details></article><article><details class="docstring" open="true"><summary id="IdentityByDescentDispersal.expected_ibd_blocks_constant_density"><a class="docstring-binding" href="#IdentityByDescentDispersal.expected_ibd_blocks_constant_density"><code>IdentityByDescentDispersal.expected_ibd_blocks_constant_density</code></a> — <span class="docstring-category">Function</span></summary><section><div><pre><code class="language-julia hljs">expected_ibd_blocks_constant_density(r::Real, D::Real, sigma::Real, L::Real, G::Real, chromosomal_edges::Bool=true, diploid::Bool=true) -&gt; Real</code></pre><p>Computes the expected density of identity-by-descent (IBD) blocks of length <code>L</code> for a model with constant population density. This function returns the expected number of IBD blocks per pair of individuals and per unit of block length.</p><ul><li><code>r</code> is the geographic distance between samples,</li><li><code>D</code> is the effective population density (diploid individuals per unit area),</li><li><code>sigma</code> is the root mean square dispersal distance per generation,</li><li><code>L</code> is the length of the IBD block (in Morgans),</li><li><code>G</code> is the total map length of the genome (in Morgans)</li></ul><p>If <code>chromosomal_edges</code> is <code>true</code> (the default), we account for chromosomal edge effects. If <code>diploid</code> is <code>true</code>, we multiply by a factor of 4 to account for the fact that each individual has two copies of each chromosome. There is a function overload that accepts a vector of <code>G</code> values and returns the aggregated expected number of IBD blocks.</p><p>We calculate it as</p><p class="math-container">\[\mathbb{E}[N_L] = \int_0^\infty \mathbb{E}[N_L^t] \,dt =
\frac{G}{8\pi D \sigma^2}
\left(\frac{r}{\sqrt{L} \sigma}\right)^2
K_2\left(\frac{\sqrt{2L} \, r}{\sigma}\right)\]</p><p>where <code>K₂</code> is the modified Bessel function of the second kind of order 2. Reference: Ringbauer, H., Coop, G., &amp; Barton, N. H. (2017). Genetics, 205(3), 1335–1351.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/ff93e5c6f8ea6a815166df0c3cf608d0d74cddd4/src/IdentityByDescentDispersal.jl#L66-L90">source</a></section></details></article><article><details class="docstring" open="true"><summary id="IdentityByDescentDispersal.expected_ibd_blocks_custom"><a class="docstring-binding" href="#IdentityByDescentDispersal.expected_ibd_blocks_custom"><code>IdentityByDescentDispersal.expected_ibd_blocks_custom</code></a> — <span class="docstring-category">Function</span></summary><section><div><pre><code class="language-julia hljs">expected_ibd_blocks_custom(r::Real, De::Function, parameters::AbstractArray, sigma::Real, L::Real, G::Real, chromosomal_edges::Bool = true, diploid::Bool = true)</code></pre><p>Computes the expected density of identity-by-descent (IBD) blocks of length <code>L</code> for a model where the effective population density is given by a custom function <code>De(t, parameters)</code>. This function returns the expected number of IBD blocks per pair of individuals and per unit of block length.</p><ul><li><code>r</code> is the geographic distance between samples,</li><li><code>De</code> is  a user-defined function that takes time <code>t</code> and a <code>parameters</code> and returns the effective population density at time <code>t</code>.</li><li><code>parameters</code> is a user-defined array of parameters that the function <code>De</code> depends on.</li><li><code>sigma</code> is the root mean square dispersal distance per generation,</li><li><code>L</code> is the length of the IBD block (in Morgans),</li><li><code>G</code> is the total map length of the genome (in Morgans),</li></ul><p>If <code>chromosomal_edges</code> is <code>true</code> (the default), we account for chromosomal edge effects. If <code>diploid</code> is <code>true</code>, we multiply by a factor of 4 to account for the fact that each individual has two copies of each chromosome. There is a function overload that accepts a vector of <code>G</code> values and returns the aggregated expected number of IBD blocks.</p><p>We calculate it as</p><p class="math-container">\[\mathbb{E}[N_L] = \int_0^\infty \mathbb{E}[N_L^t] \,dt = \int_0^\infty G  4t^2 \exp(-2Lt) \cdot \Phi(t) \,dt\]</p><p>where the integral is computed numerically using Gaussian-Legendre quadrature rules with the <code>QuadGK</code> package.</p><p>Reference: Ringbauer, H., Coop, G., &amp; Barton, N. H. (2017). Genetics, 205(3), 1335–1351.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/ff93e5c6f8ea6a815166df0c3cf608d0d74cddd4/src/IdentityByDescentDispersal.jl#L215-L238">source</a></section></details></article><article><details class="docstring" open="true"><summary id="IdentityByDescentDispersal.expected_ibd_blocks_power_density"><a class="docstring-binding" href="#IdentityByDescentDispersal.expected_ibd_blocks_power_density"><code>IdentityByDescentDispersal.expected_ibd_blocks_power_density</code></a> — <span class="docstring-category">Function</span></summary><section><div><pre><code class="language-julia hljs">expected_ibd_blocks_power_density(r::Real, D::Real, beta::Real, sigma::Real, L::Real, G::Real, chromosomal_edges::Bool = true, diploid::Bool = true)</code></pre><p>Computes the expected density of identity-by-descent (IBD) blocks of length <code>L</code> for a model with a power population density in the form of <span>$D(t) = D_0t^{-eta}$</span> This function returns the expected number of IBD blocks per pair of individuals and per unit of block length.</p><ul><li><code>r</code> is the geographic distance between samples,</li><li><code>D</code> is the effective population density (diploid individuals per unit area),</li><li><code>beta</code> is the power of the density function,</li><li><code>sigma</code> is the root mean square dispersal distance per generation,</li><li><code>L</code> is the length of the IBD block (in Morgans),</li><li><code>G</code> is the total map length of the genome (in Morgans),</li></ul><p>If <code>chromosomal_edges</code> is <code>true</code> (the default), we account for chromosomal edge effects. If <code>diploid</code> is <code>true</code>, we multiply by a factor of 4 to account for the fact that each individual has two copies of each chromosome. There is a function overload that accepts a vector of <code>G</code> values and returns the aggregated expected number of IBD blocks.</p><p>We calculate it as</p><p class="math-container">\[\mathbb{E}[N_L] = \int_0^\infty \mathbb{E}[N_L^t] \,dt =
2^{\frac{-3\beta}{2}-3}
\frac{G}{\pi D \sigma^2}
\left(\frac{r}{\sqrt{L} \sigma}\right)^{(2+\beta)}
K_{2+\beta}\left(\frac{\sqrt{2L} \, r}{\sigma}\right)\]</p><p>where <code>K₂</code> is the modified Bessel function of the second kind of order 2. Reference: Ringbauer, H., Coop, G., &amp; Barton, N. H. (2017). Genetics, 205(3), 1335–1351.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/ff93e5c6f8ea6a815166df0c3cf608d0d74cddd4/src/IdentityByDescentDispersal.jl#L138-L164">source</a></section></details></article><article><details class="docstring" open="true"><summary id="IdentityByDescentDispersal.preprocess_dataset-Tuple{DataFrames.DataFrame, DataFrames.DataFrame, AbstractVector, Real}"><a class="docstring-binding" href="#IdentityByDescentDispersal.preprocess_dataset-Tuple{DataFrames.DataFrame, DataFrames.DataFrame, AbstractVector, Real}"><code>IdentityByDescentDispersal.preprocess_dataset</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">preprocess_dataset(ibd_blocks::DataFrame, individual_distances::DataFrame, bins::AbstractVector, min_length::Real)</code></pre><p>Preprocesses the input data for identity-by-descent (IBD) block analysis.</p><ul><li><code>ibd_blocks</code>: DataFrame containing IBD blocks with columns <code>ID1</code>, <code>ID2</code>, and <code>span</code>. The <code>ID1</code> and <code>ID2</code> columns should contain the IDs of the individuals involved in the IBD block, and the <code>span</code> column should contain the length of the IBD block in Morgans.</li><li><code>individual_distances</code>: DataFrame containing distances between individuals with columns <code>ID1</code>, <code>ID2</code>, and <code>distance</code>. The <code>ID1</code> and <code>ID2</code> columns should contain the IDs of the individuals involved in the distance. Notice that the units of the estimated density and dispersal rate will match the units of the distances provided.</li><li><code>bins</code>: A vector of right bins for the IBD blocks.</li><li><code>min_length</code>: Minimum length of IBD blocks to consider in Morgans.</li></ul><p>It returns a DataFrame in &quot;long&quot; format with the following columns:</p><ul><li><code>DISTANCE</code>: The pairwise distance between individuals.</li><li><code>IBD_LEFT</code>: The left bound of the IBD length bin.</li><li><code>IBD_RIGHT</code>: The right bound of the IBD length bin.</li><li><code>IBD_MID</code>: The center of the IBD length bin.</li><li><code>NR_PAIRS</code>: The number of unique individual pairs within that distance.</li><li><code>COUNT</code>: The number of IBD blocks observed in the corresponding bin.</li><li><code>DISTANCE_INDEX</code>: The index of the distance bin.</li><li><code>IBD_INDEX</code>: The index of the IBD length bin.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/ff93e5c6f8ea6a815166df0c3cf608d0d74cddd4/src/IdentityByDescentDispersal.jl#L308-L327">source</a></section></details></article><article><details class="docstring" open="true"><summary id="IdentityByDescentDispersal.probability_coalescence-Tuple{Real, Real, Function, Real}"><a class="docstring-binding" href="#IdentityByDescentDispersal.probability_coalescence-Tuple{Real, Real, Function, Real}"><code>IdentityByDescentDispersal.probability_coalescence</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">probability_coalescence(t::Real, r::Real, De::Function, sigma::Real)</code></pre><p>Computes the probability <span>$\phi(t)$</span> that two homologous loci coalesce <span>$t$</span> generations ago according to</p><p class="math-container">\[\phi(t) = \frac{1}{2D_e(t)} \frac{1}{4\pi t \sigma^2} \exp(\frac{-r^2}{4 t \sigma^2})\]</p><p>Ringbauer, H., Coop, G. &amp; Barton, N.H. Genetics 205, 1335–1351 (2017).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/ff93e5c6f8ea6a815166df0c3cf608d0d74cddd4/src/IdentityByDescentDispersal.jl#L30-L39">source</a></section></details></article><article><details class="docstring" open="true"><summary id="IdentityByDescentDispersal.safe_adbesselk-Tuple{Any, Any}"><a class="docstring-binding" href="#IdentityByDescentDispersal.safe_adbesselk-Tuple{Any, Any}"><code>IdentityByDescentDispersal.safe_adbesselk</code></a> — <span class="docstring-category">Method</span></summary><section><div><pre><code class="language-julia hljs">x = safe_adbesselk(1, 1.0)</code></pre><p>A wrapper around the <code>adbesselk</code> to compute the modified Bessel function of the second kind of degree that returns a NaN if computation fails.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/currocam/IdentityByDescentDispersal.jl/blob/ff93e5c6f8ea6a815166df0c3cf608d0d74cddd4/src/IdentityByDescentDispersal.jl#L18-L21">source</a></section></details></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../inference/">« Inference and model evaluation</a><a class="docs-footer-nextpage" href="../references/">References »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.16.1 on <span class="colophon-date" title="Friday 13 February 2026 09:03">Friday 13 February 2026</span>. Using Julia version 1.12.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
